keras_nlp/models/falcon/falcon_tokenizer.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2024 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+
+from keras_nlp.api_export import keras_nlp_export
+from keras_nlp.models.falcon.falcon_presets import backbone_presets
+from keras_nlp.tokenizers.byte_pair_tokenizer import BytePairTokenizer
+from keras_nlp.utils.python_utils import classproperty
+
+
+@keras_nlp_export("keras_nlp.models.FalconTokenizer")
+class FalconTokenizer(BytePairTokenizer):
+    # TODO: update comments.
+    """Falcon tokenizer based on BytePairTokenizer.
+
+    This tokenizer class will tokenize raw strings into integer sequences and
+    is based on `keras_nlp.tokenizers.BytePairTokenizer`. Unlike the
+    underlying tokenizer, it will check for all special tokens needed by Falcon
+    models and provides a `from_preset()` method to automatically download
+    a matching vocabulary for a Falcon preset.
+
+    This tokenizer does not provide truncation or padding of inputs.
+
+    If input is a batch of strings (rank > 0), the layer will output a
+    `tf.RaggedTensor` where the last dimension of the output is ragged.
+
+    If input is a scalar string (rank == 0), the layer will output a dense
+    `tf.Tensor` with static shape `[None]`.
+
+    Args:
+        vocabulary: string or dict, maps token to integer ids. If it is a
+            string, it should be the file path to a json file.
+        merges: string or list, contains the merge rule. If it is a string,
+            it should be the file path to merge rules. The merge rule file
+            should have one merge rule per line. Every merge rule contains
+            merge entities separated by a space.
+
+    Examples:
+
+    ```python
+    # Unbatched input.
+    tokenizer = keras_nlp.models.FalconTokenizer.from_preset("falcon_rw_1b")
+    tokenizer("The quick brown fox jumped.")
+
+    # Batched input.
+    tokenizer(["The quick brown fox jumped.", "The fox slept."])
+
+    # Detokenization.
+    tokenizer.detokenize(tokenizer("The quick brown fox jumped."))
+
+    # Custom vocabulary.
+    vocab = {"<|endoftext|>": 0, "a": 4, "Ġquick": 5, "Ġfox": 6}
+    merges = ["Ġ q", "u i", "c k", "ui ck", "Ġq uick"]
+    merges += ["Ġ f", "o x", "Ġf ox"]
+    tokenizer = keras_nlp.models.FalconTokenizer(vocabulary=vocab, merges=merges)
+    tokenizer("a quick fox.")
+    ```
+    """
+
+    def __init__(
```

Comment:
General question, but what is up with the 7b tokenizer? It looks like it is still basically just BPE, but with extra special tokens? https://huggingface.co/tiiuae/falcon-7b-instruct/raw/main/tokenizer.json

Maybe we can pull the vocab and merges out of this json, so we can handle them normally, and tackle the rest of the weirdness in code?

Reply:
I think it's not just special token differences. They have different vocab sizes: 1b has `50256` tokens in total while 7b has `65023`  (there are only 10 extra special tokens in 7b).

Reply:
Yeah, new tokenizer vocab for sure, but I was hoping we could avoid a whole new file format.

It seems like they are still fundamentally BPE, with a different vocab and more special tokens right? If we can still save this as a `tokenizer.json` + `assets/tokenizer/merges.txt` + `assets/tokenizer/vocab.json` that seems ideal to me. But we could also write a custom loader for Falcon's bespoke `tokenizer.json` format if we thing that's better.

Reply:
I misunderstood what you said.

Do you mean that since they are basically using BPE, we can skip creating `FalconTokenizer` and use BPE directly?


> I was hoping we could avoid a whole new file format.

Could you explain what you mean by "a new file format"? What's the new file format I'm creating? :D

Reply:
Sorry! I am being unclear. Everything you have looks good for the 1b model. I am asking about/trying to think through an upcoming problem with the 7b falcon models.

Take a look at:
- https://huggingface.co/tiiuae/falcon-rw-1b/tree/main
- https://huggingface.co/tiiuae/falcon-7b/tree/main

The 7b tokenizer assets are different, there is no merges.txt or vocab.json. There is just one weird `tokenizer.json` that combines the two. We don't have any code that will allow reading that bespoke `tokenizer.json` today, so we could either extrac a merges and vocab from it, support loading it directly with new parsing code, or something else.

Does that clarify or not really?

Reply:
And to be clear, we should have a `FalconTokenizer` for sure. The question I have is just whether it can be a "simple subclass" of the `BytePairEncoding` tokenizer, or whether we need custom json parsing code after we also convert the 7b models.

Reply:
Oh, I see! Thanks for the clarification!

I agree that it would be better to just extract `vocab` and `merges` from their format and load it like other models as there isn't any other model that has this format.

Reply:
I am not sure if I understand what you arguing right but I believe it's about special tokens that the 7 billion model have and how we can save them properly.
If that is the case so we can add `special_tokens` arg for FalconTokenizer, something like `WhisperTokenizer`, but instead it can be only a list because they are already included in the vocabulary, and while initialization, we pass them with `<|endoftext|>` in `usplittable_tokens` arg while initializing the super class .
And for the conversion script, while converting the tokenizer, we can check for `hf_tokenizer["added_tokens"]` list if it has any added tokens other than the `<|endoftext|>` token and pass them to the `FalconTokenizer` as `special_tokens`. and then of course we need to update the config to contain the special tokens. So we will have only `tokenizer.json` + `assets/tokenizer/merges.txt` + `assets/tokenizer/vocab.json`, but with the config in `tokenizer.json` having a list of `special_tokens`.